### PR TITLE
[ONNX] Add complex constant support

### LINKED ISSUE #137864: [ONNX] Handle complex python constants
### 🐛 Describe the bug

```
TypeError: Constant input '[None, _unsafe_view defined in (%_unsafe_view : Int(7168) = onnx::Reshape[allowzero=0](%clone, %_val_33)
)]' of type '<class 'list'>' is not supported
```

[this](https://github.com/tuna2134/f5-tts/blob/master/onnx/convert_vocos.py) is source code

[sarief](https://github.com/tuna2134/f5-tts/blob/master/report_dynamo_export.sarif)

### Versions
```
Collecting environment information...
PyTorch version: 2.6.0.dev20241013+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: (Ubuntu 13.2.0-23ubuntu4) 13.2.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.4 (main, Aug  7 2024, 12:31:38) [GCC 13.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-45-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        40 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               8
On-line CPU(s) list:                  0-7
Vendor ID:                            GenuineIntel
Model name:                           Common KVM processor
CPU family:                           15
Model:                                6
Thread(s) per core:                   1
Core(s) per socket:                   8
Socket(s):                            1
Stepping:                             1
BogoMIPS:                             4794.41
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx lm constant_tsc nopl xtopology cpuid tsc_known_freq pni cx16 x2apic hypervisor lahf_lm cpuid_fault pti
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            256 KiB (8 instances)
L1i cache:                            256 KiB (8 instances)
L2 cache:                             32 MiB (8 instances)
L3 cache:                             16 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-7
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          KVM: Mitigation: VMX unsupported
Vulnerability L1tf:                   Mitigation; PTE Inversion
Vulnerability Mds:                    Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:               Mitigation; PTI
Vulnerability Mmio stale data:        Unknown: No mitigations
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Vulnerable
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Retpoline
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.1.2
[pip3] onnx==1.17.0
[pip3] onnxscript==0.1.0.dev20241013
[pip3] torch==2.6.0.dev20241013+cpu
[pip3] torchaudio==2.5.0.dev20241013+cpu
```

cc @ezyang @chauhang @penguinwu @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

### PR DESCRIPTION
**Pull request by titaiwangms**
Transform complex python constant to float representation as well, like what we have with tensors.

PS: I find it's not reasonable to add "complex->float" in IR side, so I put it here.

### CODE REVIEW

File: torch/onnx/_internal/exporter/_building.py [hunk 1]
```diff
@@ -334,6 +336,11 @@ def _process_python_constants(
 
         dtype = _determine_input_dtype(param, arg, type_binding)
 
+        # float representation of complex numbers
+        if isinstance(arg, complex):
+            # Convert the complex number to a float
+            arg = [arg.real, arg.imag]
```
Comment (justinchuby) [inline torch/onnx/_internal/exporter/_building.py:342]: Should we handle this in _get_or_create_constant? Just thinking

**Comment by titaiwangms** [comment]
If I try with the following repro to prove the fix:

```python
import torch

class MulModule(torch.nn.Module):
    def forward(self, x, y):
        return torch.ops.aten.mul(x, y)

# Example usage with complex inputs
x = torch.tensor([[1.0 + 2.0j, 3.0 + 4.0j], 
                           [5.0 + 6.0j, 7.0 + 8.0j]], dtype=torch.complex64)

# Example 1: Non-tensor input (scalar)
y = 2 + 3j

onnx_program = torch.onnx.export(MulModule(), (x, y,), "slice.onnx", dynamo=True, report=True)
```

I get 
```
[torch.onnx] Obtain model graph for `MulModule()` with `torch.export.export`...
[torch.onnx] Obtain model graph for `MulModule()` with `torch.export.export`... ❌
[torch.onnx] Obtain model graph for `MulModule()` with `torch.export.export(..., strict=False)`...
[torch.onnx] Obtain model graph for `MulModule()` with `torch.export.export(..., strict=False)`... ❌
[torch.onnx] Obtain model graph for `MulModule()` with Torch Script...
[torch.onnx] Obtain model graph for `MulModule()` with Torch Script... ❌
[torch.onnx] Obtain model graph for `MulModule()` with internal Dynamo apis...
[torch.onnx] Obtain model graph for `MulModule()` with internal Dynamo apis... ❌
Traceback (most recent call last):
  File "/home/titaiwang/pytorch/torch/onnx/_internal/exporter/_capture_strategies.py", line 110, in __call__
    exported_program = self._capture(model, args, kwargs, dynamic_shapes)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/titaiwang/pytorch/torch/onnx/_internal/exporter/_capture_strategies.py", line 145, in _capture
    return torch.export.export(
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/titaiwang/pytorch/torch/export/__init__.py", line 368, in export
    return _export(
           ^^^^^^^^
  File "/home/titaiwang/pytorch/torch/export/_trace.py", line 1018, in wrapper
    raise e
  File "/home/titaiwang/pytorch/torch/export/_trace.py", line 991, in wrapper
    ep = fn(*args, **kwargs)
         ^^^^^^^^^^^^^^^^^^^
  File "/home/titaiwang/pytorch/torch/export/exported_program.py", line 122, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/titaiwang/pytorch/torch/export/_trace.py", line 1974, in _export
    export_artifact = export_func(  # type: ignore[operator]
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/titaiwang/pytorch/torch/export/_trace.py", line 1238, in _strict_export
    return _strict_export_lower_to_aten_ir(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/titaiwang/pytorch/torch/export/_trace.py", line 1347, in _strict_export_lower_to_aten_ir
    aten_export_artifact = lower_to_aten_callback(
                           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/titaiwang/pytorch/torch/export/_trace.py", line 647, in _export_to_aten_ir
    gm, graph_signature = transform(aot_export_module)(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/titaiwang/pytorch/torch/_functorch/aot_autograd.py", line 1262, in aot_export_module
    fx_g, metadata, in_spec, out_spec = _aot_export_function(
                                        ^^^^^^^^^^^^^^^^^^^^^
  File "/home/titaiwang/pytorch/torch/_functorch/aot_autograd.py", line 1497, in _aot_export_function
    fx_g, meta = create_aot_dispatcher_function(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/titaiwang/pytorch/torch/_functorch/aot_autograd.py", line 524, in create_aot_dispatcher_function
    return _create_aot_dispatcher_function(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/titaiwang/pytorch/torch/_functorch/aot_autograd.py", line 625, in _create_aot_dispatcher_function
    fw_metadata = run_functionalized_fw_and_collect_metadata(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/titaiwang/pytorch/torch/_functorch/_aot_autograd/collect_metadata_analysis.py", line 171, in inner
    assert all(isinstance(a, tuple(KNOWN_TYPES)) for a in flat_args)
AssertionError
```

**Comment by justinchuby** [comment]
I see. Just move it to inside the forward function?

File: torch/onnx/_internal/exporter/_building.py (hunk 1 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_building.py]: Done

**Comment by titaiwangms** [comment]
> I see. Just move it to inside the forward function?

Thanks! A test can be added!

**Comment by titaiwangms** [comment]
@justinchuby PTAL

File: test/onnx/exporter/test_small_models_e2e.py [hunk 2]
```diff
@@ -33,6 +33,20 @@ def forward(self, query, key, value):
         )
         onnx_testing.assert_onnx_program(onnx_program, atol=1e-3, rtol=1)
 
+    def test_exporter_supports_constant_complex(self):
```
Comment (justinchuby) [inline test/onnx/exporter/test_small_models_e2e.py:36]:
```suggestion
    def test_constant_complex(self):
```

File: torch/onnx/_internal/exporter/_building.py [hunk 3]
```diff
@@ -261,9 +263,15 @@ def _get_or_create_constant(
     dtype: ir.DataType,
     opset: onnxscript.values.Opset,
 ) -> ir.Value:
+    # float representation of complex numbers
+    if isinstance(arg, complex):
+        # Convert the complex number to a float
+        arg = [arg.real, arg.imag]
```
Comment (justinchuby) [inline torch/onnx/_internal/exporter/_building.py:269]:
```suggestion
        arg = (arg.real, arg.imag)
```

simpler

**Review by justinchuby — APPROVED** [review]

**Pull request closed without merging**